### PR TITLE
SDIT-1350 Report on activity migration suspended allocations

### PR DESCRIPTION
**Pull request by mikehalmamoj**
The **Found suspended allocations** message now appears in the migration preview checks:
image

And clicking on the **Copy** button adds the following to the Clipboard:
```
Activity Description, Activity ID, Prisoner Number,
AM & PM A Hotplate Assistant, 81641, G8044VV,
AM & PM B Wing Cleaner, 81672, G0065UT,
AM & PM B Wing Cleaner, 81672, G8271VD,
AM & PM C Wing Cleaner, 81674, G0803GG,
AM & PM C Wing Cleaner, 81674, G6458GA,
AM & PM C Wing Cleaner, 81674, G9918GE,
AM & PM G Hotplate Assistant, 81650, G2067VO,
AM & PM G Wing OAP, 157860, G4110UF,
AM & PM PIDS A Wing, 160355, G4394VJ,
...etc
```

### CODE REVIEW

File: server/routes/activitiesMigration/activitiesMigrationController.ts [hunk 1]
```diff
@@ -128,6 +128,14 @@ export default class ActivitiesMigrationController {
...
         return null
       }),
+
+      this.nomisMigrationService.findActivitiesSuspendedAllocations(filter, context(res)).catch(error => {
+        errors.push({
+          text: `Failed to find suspended allocations due to error: ${error.message}`,
+          href: '',
+        })
+        return []
+      }),
```
Comment (mikehalmamoj) [inline server/routes/activitiesMigration/activitiesMigrationController.ts:138]: Get suspended allocations

File: server/routes/activitiesMigration/activitiesMigrationController.ts [hunk 2]
```diff
@@ -151,10 +160,24 @@ export default class ActivitiesMigrationController {
...
+    const body = allocations
+      .map(
+        (allocation: FindSuspendedAllocationsResponse) =>
+          `${allocation.courseActivityDescription}, ${allocation.courseActivityId}, ${allocation.offenderNo},`,
+      )
+      .sort()
+    body.unshift(`Activity Description, Activity ID, Prisoner Number,`)
+    return body
+  }
+
```
Comment (mikehalmamoj) [inline server/routes/activitiesMigration/activitiesMigrationController.ts:180]: Pass suspended allocations to the front end in csv format

File: server/views/pages/activities/startActivitiesMigrationPreview.njk [hunk 3]
```diff
@@ -62,6 +62,13 @@
                                     DPS has no slot times configured for prison {{ form.prisonId }}.
                                 </p>
                             {% endif %}
+                            {% if form.suspendedAllocations|length %}
+                                <p class="govuk-notification-banner__heading" id="nomisSuspendedAllocations">
+                                    Found suspended allocations in NOMIS for prison {{ form.prisonId }}.
+                                    <a href="#" id='copy-suspended-link'">Copy</a>
+                                    <textarea class="govuk-visually-hidden" id="copy-suspended-text" rows="{{ form.suspendedAllocations|length }}" cols="200" readonly>{{ form.suspendedAllocations|join("\n") }}</textarea>
+                                </p>
```
Comment (mikehalmamoj) [inline server/views/pages/activities/startActivitiesMigrationPreview.njk:70]: Save suspended allocations csv in a hidden text area with a related "Copy" link

File: assets/js/page-enhancements.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      const copyTextArea = document.getElementById(`${prefix}-text`)
+      if (copyLink) {
+        if (navigator.clipboard && copyTextArea) {
+          copyLink.addEventListener('click', async () => {
+            await copyText(`${prefix}-text`)
+          })
+        } else {
+          copyLink.classList.add('govuk-visually-hidden')
+        }
+      }
```
Comment (mikehalmamoj) [inline assets/js/page-enhancements.js:31]: Only add the click event to the copy link if the clipboard is available

File: assets/js/page-enhancements.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ */
+window.pageEnhancements = (($, document) => {
+  const copyTextToClipboard = () => {
+    if (!document.getElementById('startActivitiesMigrationPreviewPage')) {
+      return
+    }
+    async function copyText(textAreaId) {
+      const text = document.getElementById(textAreaId).value
+      await navigator.clipboard.writeText(text)
+    }
```
Comment (mikehalmamoj) [inline assets/js/page-enhancements.js:18]: Grab the hidden text area contents and write htem to the clipboard

File: assets/js/page-enhancements.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ * If the user does not have JS enabled, or the code within this script does not run for some reason
+ * the application should fall back to standard HTML functionality and should continue to operate.
+ *
+ * No code should be written here that renders a page non-functional if JS is not enabled or the script does not run/fails.
+ */
+window.pageEnhancements = (($, document) => {
+  const copyTextToClipboard = () => {
+    if (!document.getElementById('startActivitiesMigrationPreviewPage')) {
+      return
+    }
```
Comment (mikehalmamoj) [inline assets/js/page-enhancements.js:14]: Do nothing if on the wrong page

**Review by andymarke — APPROVED** [review]
lgtm

**Pull request merged**